File: medBE/src/main/java/com/SxxM/med/analysis/service/GptService.java
package com.SxxM.med.analysis.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.stereotype.Service;
import org.springframework.web.reactive.function.client.WebClient;

import java.util.List;
import java.util.Map;

@Service
@RequiredArgsConstructor
@Slf4j
public class GptService {
    
    private final WebClient.Builder webClientBuilder;
    private final ObjectMapper objectMapper;
    
    @Value("${gpt.api.key}")
    private String apiKey;
    
    @Value("${gpt.api.url:https://api.openai.com/v1/chat/completions}")
    private String apiUrl;
    
    @Value("${gpt.model:gpt-4}")
    private String model;
    
    private WebClient getWebClient() {
        return webClientBuilder
                .baseUrl(apiUrl)
                .defaultHeader(HttpHeaders.AUTHORIZATION, "Bearer " + apiKey)
                .defaultHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE)
                .build();
    }
    
    public <T> T analyzeWithGpt(String prompt, Class<T> responseClass) {
        try {
            Map<String, Object> requestBody = Map.of(
                    "model", model,
                    "messages", List.of(
                            Map.of("role", "system", "content", "You are a medical assistant. Always respond in valid JSON format only."),
                            Map.of("role", "user", "content", prompt)
                    ),
                    "temperature", 0.3,
                    "response_format", Map.of("type", "json_object")
            );
            
            String response = getWebClient()
                    .post()
                    .bodyValue(requestBody)
                    .retrieve()
                    .bodyToMono(String.class)
                    .block();
            
            if (response == null) {
                throw new RuntimeException("GPT API 응답이 null입니다");
            }
            
            Map<String, Object> responseMap = objectMapper.readValue(response, Map.class);
            List<Map<String, Object>> choices = (List<Map<String, Object>>) responseMap.get("choices");
            if (choices == null || choices.isEmpty()) {
                throw new RuntimeException("GPT API 응답에 choices가 없습니다");
            }
            
            Map<String, Object> message = (Map<String, Object>) choices.get(0).get("message");
            String content = (String) message.get("content");
            
            return objectMapper.readValue(content, responseClass);
        } catch (Exception e) {
            log.error("GPT API 호출 중 오류 발생", e);
            throw new RuntimeException("GPT 분석 중 오류가 발생했습니다: " + e.getMessage(), e);
        }
    }
    
    public String analyzeWithGptString(String prompt) {
        try {
            Map<String, Object> requestBody = Map.of(
                    "model", model,
                    "messages", List.of(
                            Map.of("role", "system", "content", "You are a medical assistant. Provide detailed analysis."),
                            Map.of("role", "user", "content", prompt)
                    ),
                    "temperature", 0.3
            );
            
            String response = getWebClient()
                    .post()
                    .bodyValue(requestBody)
                    .retrieve()
                    .bodyToMono(String.class)
                    .block();
            
            if (response == null) {
                throw new RuntimeException("GPT API 응답이 null입니다");
            }
            
            Map<String, Object> responseMap = objectMapper.readValue(response, Map.class);
            List<Map<String, Object>> choices = (List<Map<String, Object>>) responseMap.get("choices");
            if (choices == null || choices.isEmpty()) {
                throw new RuntimeException("GPT API 응답에 choices가 없습니다");
            }
            
            Map<String, Object> message = (Map<String, Object>) choices.get(0).get("message");
            return (String) message.get("content");
        } catch (Exception e) {
            log.error("GPT API 호출 중 오류 발생", e);
            throw new RuntimeException("GPT 분석 중 오류가 발생했습니다: " + e.getMessage(), e);
        }
    }
}

File: medPY/app/services/gpt_service.py
import os
import json
import httpx
from typing import Dict, Any, Optional
from dotenv import load_dotenv

load_dotenv()

class GptService:
    def __init__(self):
        self.api_key = os.getenv("OPENAI_API_KEY", "")
        self.api_url = os.getenv("OPENAI_API_URL", "https://api.openai.com/v1/chat/completions")
        self.model = os.getenv("GPT_MODEL", "gpt-4o-mini")
    
    async def analyze_with_gpt(self, prompt: str) -> Dict[str, Any]:
        """
        GPT API를 호출하여 분석을 수행합니다.
        """
        if not self.api_key:
            raise ValueError("OPENAI_API_KEY 환경변수가 설정되지 않았습니다")
        
        headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json"
        }
        
        payload = {
            "model": self.model,
            "messages": [
                {"role": "system", "content": "You are a medical assistant. Always respond in valid JSON format only."},
                {"role": "user", "content": prompt}
            ],
            "temperature": 0.3,
            "response_format": {"type": "json_object"}
        }
        
        async with httpx.AsyncClient(timeout=60.0) as client:
            response = await client.post(self.api_url, headers=headers, json=payload)
            response.raise_for_status()
            
            result = response.json()
            
            if "choices" not in result or not result["choices"]:
                raise ValueError("GPT API 응답에 choices가 없습니다")
            
            content = result["choices"][0]["message"]["content"]
            
            # JSON 파싱
            try:
                return json.loads(content)
            except json.JSONDecodeError as e:
                raise ValueError(f"GPT 응답을 JSON으로 파싱할 수 없습니다: {e}, 응답: {content}")

